### ISSUE TITLE
Nesting 'masked' wrappers gives a confusing error

### ISSUE THREAD

**Issue by tomweingarten** [deepmind/optax#271]
The _masked_ wrapper requires the model parameters to be input as a pytree. It flattens that pytree before calling init on its inner optimizer. So if you wrap masked around a optimizer that also uses masked, like lamb, optax throws a mismatched type error. It expects to receive a pytree, but instead receives a flat list of parameters. In the short term, I recommend we throw a useful error when this occurs, since it's relatively easy to mitigate this through the use of chain. In the long term I recommend rebuilding the tree in masked before passing it to the inner optimizer.

**Comment by n2cholas** [comment]
Thanks for raising this issue, @tomweingarten! Does [this](https://github.com/n2cholas/optax/blob/nested_masked/optax/_src/wrappers.py#L328-L344) implementation fix your nested masked issue (install via `pip install --upgrade git+https://github.com/n2cholas/optax@nested_masked`)? If so, you have a minimal version of the code that failed before? I can turn it into a test case. 

This version preserves the pytree structure by replacing masked values with `None`. The original implementation was unnecessarily complex with the flattening because I did not realize `None` was a valid dummy value (since it is considered an empty pytree). 

However, this change will break user code that relies on the flattened state. @mkunesch @mtthss, any advice on whether it is reasonable to merge this change (perhaps in the next major version change)?

**Comment by mkunesch** [comment]
Hi! Thanks a lot @tomweingarten for reporting and @n2cholas for suggesting a fix. The fix looks great to me!

I personally think it's unlikely that there are many users who rely on the internal state of the optimizer being flattened so I think we can just make the change. Also, it looks like the `masked` wrapper is covered well by tests (apart perhaps from a test with a callable mask?) so it seems like a safe change to make.

@n2cholas : do you have time to file a PR with this?

**Comment by mkunesch** [comment]
Yes, thanks for cc'ing! I also just clarified my comment above: it's the parameters which are flattened by the current implementation of `masked` before being passed to the inner gradient transformations, the tree structure of the optimizer state itself is not affected.

**Comment by andsteing** [comment]
I don't think we depend anywhere on internal optimizer's state being flattened.

Nice to see this change, being able to use `optax.masked()` in a nested way can facilitate the design of composite gradient transformations...

**Comment by n2cholas** [comment]
> @n2cholas : do you have time to file a PR with this?

Yep, will do this week!

> Also, it looks like the `masked` wrapper is covered well by tests (apart perhaps from a test with a callable mask?) so it seems like a safe change to make.

I'll add a callable test and a nested test as well.

**Comment by mkunesch** [comment]
Amazing, thank you so much @n2cholas!
